micropython/utop/utop.py

```diff
@@ -0,0 +1,108 @@
+import micropython
+import time
+import _thread
+
+try:
+    import esp32
+except ImportError:
+    esp32 = None
+
+
+def top(update_interval_ms=1000, timeout_ms=None, thread_names={}):
+    time_start = time.ticks_ms()
+    previous_total_runtime = 0
+    previous_task_runtimes = {}
+    previous_line_count = 0
+    esp32_task_state_names = ("running", "ready", "blocked", "suspended", "deleted", "invalid")
+
+    while timeout_ms is None or abs(time.ticks_diff(time.ticks_ms(), time_start)) < timeout_ms:
+        if previous_line_count > 0:
+            print("\x1b[{}A".format(previous_line_count), end="")
+        line_count = 0
+
+        if esp32 is not None:
+            if not hasattr(esp32, "idf_task_stats"):
+                print(
+                    "INFO: esp32.idf_task_stats() is not available, cannot list active tasks.\x1b[K"
+                )
+                line_count += 1
+            else:
+                print("   CPU%  CORE  PRIORITY  STATE      STACKWATERMARK  NAME\x1b[K")
+                line_count += 1
+
+                total_runtime, tasks = esp32.idf_task_stats()
+                current_thread_id = _thread.get_ident()
+                tasks.sort(key=lambda t: t[0])
+                for (
+                    task_id,
+                    task_name,
+                    task_state,
+                    task_priority,
+                    task_runtime,
+                    task_stackhighwatermark,
+                    task_coreid,
+                ) in tasks:
+                    task_runtime_percentage = "-"
+                    if previous_total_runtime != 0 and task_id in previous_task_runtimes:
+                        task_runtime_percentage = "{:.2f}%".format(
+                            100
+                            * time.ticks_diff(task_runtime, previous_task_runtimes[task_id])
+                            / time.ticks_diff(total_runtime, previous_total_runtime)
```

Comment:
Is `ticks_diff()` the correct thing to use here?  AFAIK, these times are from FreeRTOS and are not in the correct wrap-around-range to use with `ticks_diff`.

Reply:
You're right, indeed that should just be a plain 32bit unsigned wraparound.